drivers/Watchdog.cpp

```diff
@@ -18,48 +18,103 @@
 
 #include "Watchdog.h"
 
-namespace mbed {
+namespace mbed
+{
+
+Watchdog *Watchdog::first;
 
-watchdog_status_t Watchdog::start(const uint32_t timeout)
+bool Watchdog::add_to_list()
 {
-    if (timeout == 0) {
-        return WATCHDOG_STATUS_INVALID_ARGUMENT;
+    Watchdog *node;
+    node = first;
+    if(first == NULL) {
+        first = this;
+        this->next=NULL;
+    } else {
+        while(node->next != NULL) {
+            node = node->next;
+            if(node == this) {
+                return false;
+            }
+        }
+        this->next = NULL;
+        node->next = this;
     }
+    return true;
+}
 
-    if (timeout > max_timeout()) {
-        return WATCHDOG_STATUS_INVALID_ARGUMENT;
+void Watchdog::start()
+{
+    bool msts = false;
+    MBED_ASSERT(!_is_initialized);
+    core_util_critical_section_enter();
+    msts = add_to_list();
+    if(msts) {
+        _is_initialized = true;
     }
-
-    watchdog_config_t config;
-    config.timeout_ms = timeout;
-
-    return hal_watchdog_init(&config);
+    core_util_critical_section_exit();
 }
 
 
 void Watchdog::kick()
 {
-    hal_watchdog_kick();
+    MBED_ASSERT(_is_initialized);
+    core_util_critical_section_enter();
+    _current_count = 0;
+    core_util_critical_section_exit();
 }
 
-
-watchdog_status_t Watchdog::stop()
+void Watchdog::stop()
 {
-    return hal_watchdog_stop();
+    bool msts = false;
+    MBED_ASSERT(_is_initialized);
+    core_util_critical_section_enter();
+    msts = remove_from_list();
+    if(msts) {
+        _is_initialized = false;
+    }
+    core_util_critical_section_exit();
 }
 
-
-uint32_t Watchdog::reload_value() const
+bool Watchdog::remove_from_list()
 {
-    return hal_watchdog_get_reload_value();
+    Watchdog *cur_ptr = first,
+              *prev_ptr = NULL;
+    while(cur_ptr != NULL) {
+        if(cur_ptr == this) {
+            if(cur_ptr == first) {
+                prev_ptr = first;
+                first = cur_ptr->next;
+                prev_ptr->next = NULL;
+            } else {
+                prev_ptr->next = cur_ptr->next;
+                cur_ptr->next = NULL;
+            }
+            return true;
+        } else {
+            prev_ptr = cur_ptr;
+            cur_ptr = cur_ptr->next;
+        }
+    }
+    return false;
 }
 
-
-uint32_t Watchdog::max_timeout()
+void Watchdog::is_alive()
 {
-    const watchdog_features_t features = hal_watchdog_get_platform_features();
+    Watchdog *cur_ptr =  first;
+    while(cur_ptr != NULL) {
+        if(cur_ptr->_current_count > cur_ptr->_max_timeout) {
+            system_reset();
+        } else {
+            cur_ptr->_current_count++;
+            cur_ptr = cur_ptr->next;
+        }
+    }
+}
 
-    return features.max_timeout;
+Watchdog::~Watchdog()
+{
+    stop();
```

Comment:
This should check `_is_initialized` so you don't get an assert if an object is destroyed without a call to `start()`.